python/pyspark/sql/connect/udtf.py

```diff
@@ -0,0 +1,206 @@
+
+    # Create a regular Python UDTF and check for invalid handler class.
+    regular_udtf = _create_udtf(cls, returnType, name, PythonEvalType.SQL_TABLE_UDF, deterministic)
+
+    if not arrow_enabled:
+        return regular_udtf
+
+    from pyspark.sql.pandas.utils import (
+        require_minimum_pandas_version,
+        require_minimum_pyarrow_version,
+    )
+
+    try:
+        require_minimum_pandas_version()
+        require_minimum_pyarrow_version()
+    except ImportError as e:
+        warnings.warn(
+            f"Arrow optimization for Python UDTFs cannot be enabled: {str(e)}. "
+            f"Falling back to using regular Python UDTFs.",
+            UserWarning,
+        )
+        return regular_udtf
+
+    from pyspark.sql.udtf import _vectorize_udtf
+
+    vectorized_udtf = _vectorize_udtf(cls)
+    return _create_udtf(
+        vectorized_udtf, returnType, name, PythonEvalType.SQL_ARROW_TABLE_UDF, deterministic
+    )
+
+
+class UserDefinedTableFunction:
+    """
+    User defined function in Python
+
+    Notes
+    -----
+    The constructor of this class is not supposed to be directly called.
+    Use :meth:`pyspark.sql.functions.udtf` to create this instance.
+    """
+
+    def __init__(
+        self,
+        func: Type,
+        returnType: Union[StructType, str],
+        name: Optional[str] = None,
+        evalType: int = PythonEvalType.SQL_TABLE_UDF,
+        deterministic: bool = True,
+    ) -> None:
+        self.func = func
+        self.returnType: DataType = (
+            UnparsedDataType(returnType) if isinstance(returnType, str) else returnType
+        )
+        self._name = name or func.__name__
+        self.evalType = evalType
+        self.deterministic = deterministic
+
+        _validate_udtf(self)
+
+    def _build_common_inline_user_defined_table_function(
+        self, *cols: "ColumnOrName"
+    ) -> CommonInlineUserDefinedTableFunction:
+        arg_cols = [
+            col if isinstance(col, Column) else Column(ColumnReference(col)) for col in cols
+        ]
+        arg_exprs = [col._expr for col in arg_cols]
+
+        udtf = PythonUDTF(
+            func=self.func,
+            return_type=self.returnType,
+            eval_type=self.evalType,
+            python_ver=get_python_ver(),
+        )
+        return CommonInlineUserDefinedTableFunction(
+            function_name=self._name,
+            function=udtf,
+            deterministic=self.deterministic,
+            arguments=arg_exprs,
+        )
+
+    def __call__(self, *cols: "ColumnOrName") -> "DataFrame":
+        from pyspark.sql.connect.dataframe import DataFrame
+        from pyspark.sql.connect.session import _active_spark_session
+
+        if _active_spark_session is None:
+            # TODO: can we create a Spark session here?
+            raise PySparkRuntimeError(
+                "An active SparkSession is required for "
+                "executing a Python user-defined table function."
+            )
```

Comment:
@HyukjinKwon @zhengruifeng Given that we don't have `SparkSession._getActiveSessionOrCreate()` in Spark Connect, what's the best approach to creating a Spark Connect session when the active Spark session is None? Or, should we not recommend this and instead throw an exception when the Spark Connect session is None?

Reply:
currently, we use `pyspark.sql.connect.session._active_spark_session` in ml and pandas

https://github.com/apache/spark/blob/10751dc285c5c639e3343a8abc26857407522822/python/pyspark/ml/util.py#L756-L758

https://github.com/apache/spark/blob/b97ce8b9a99c570fc57dec967e7e9db3d115c1db/python/pyspark/pandas/utils.py#L484-L486

Reply:
oh, we don't have `getActiveSessionOrCreate` now